lib/zendesk_apps_tools/api_connection.rb

```diff
@@ -1,11 +1,26 @@
 module ZendeskAppsTools
   module APIConnection
-    FULL_URL     = /https?:\/\//
-    URL_TEMPLATE = 'https://%s.zendesk.com/'
+    DEFAULT_URL_TEMPLATE = 'https://%s.zendesk.com/'
+    # taken from zendesk/lib/vars.rb
+    SUBDOMAIN_VALIDATION_PATTERN = /^[a-z0-9][a-z0-9\-]{1,}[a-z0-9]$/i
+    ZENDESK_URL_VALIDATION_PATTERN = /^(https?):\/\/[a-z0-9]+(([\.]|[\-]{1,2})[a-z0-9]+)*\.([a-z]{2,16}|[0-9]{1,3})((:[0-9]{1,5})?(\/?|\/.*))?$/ix
+    EMAIL_REGEX  = /^([\w\.%\+\-]+)@([\w\-]+\.)+([\w]{2,})$/i
+
+    EMAIL_ERROR_MSG = 'Please enter a valid email address.'
+    PROMPT_FOR_URL  = 'Enter your Zendesk subdomain or full URL (including protocol):'
```

Comment:
Good idea!

Reply:
I had a thought about this.

If I created 
```
(constants declaration)
PROMPT_FOR_USERNAME = ''Enter your username:''
PROMPT_FOR_PASSWORD = ''Enter your password:''

(calling for the constants)
@username  ||= cache.fetch('username', @subdomain) || get_value_from_stdin(PROMPT_FOR_USERNAME)
@password  ||= cache.fetch('password', @subdomain) || get_password_from_stdin(PROMPT_FOR_PASSWORD')
```
It definitely promotes code consistency 👍 💯 . But I will be adding 2 additional lines of code that may not be necessary. I put it in a constant because the code was extending horizontally which made reading a bit harder (with horizontal scrolling), as such. 

```
@subdomain ||= cache.fetch('subdomain') || get_value_from_stdin('Enter your Zendesk subdomain or full URL (including protocol):')
say_error_and_exit URL_ERROR_MSG unless valid_subdomain? || valid_full_url?
```

I did think of a few more ways to expend of this thought, though. What if I put it in a common method? Something like this
```
def message_prompt_for_input(attribute)
  "Enter your #{attribute}"
end

(and call)
get_value_from_stdin(message_prompt_for_input('username'))
get_value_from_stdin(message_prompt_for_input('password'))

(or even extract the entire line, then call)
def fetch_cache_or_user_prompt(attribute)
  cache.fetch(attribute, @subdomain) || get_value_from_stdin("Enter your #{attribute}:")
end

@username  ||= fetch_cache_or_user_prompt(username)
@password  ||= fetch_cache_or_user_prompt(password)
```

But this would steer very far away from what we are trying to achieve here in terms of complexity. I think as for now it would be alright to leave them as they are until we have to change the strings and it gets super long. :)